Marking Heartbeat as healthy is too late

When a unhealthy connection becomes healthy again by succeeding a heartbeat we are sending an request to clear listeners. Since we do not mark the connection as healthy again before this request, the request assumes that connection is still unresponsive and throws the following exception:

Client Log
Jul 17, 2014 5:24:27 PM com.hazelcast.client.connection.nio.ClientConnection
WARNING: Clearing listeners upon recovering from heart-attack failed
java.util.concurrent.ExecutionException: com.hazelcast.spi.exception.TargetNotMemberException: Address : Address[10.174.36.244]:33002
at com.hazelcast.client.spi.impl.ClientCallFuture.resolveResponse(ClientCallFuture.java:195)
at com.hazelcast.client.spi.impl.ClientCallFuture.get(ClientCallFuture.java:126)
at com.hazelcast.client.connection.nio.ClientConnection.heartBeatingSucceed(ClientConnection.java:397)
at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$HeartBeat.run(ClientConnectionManagerImpl.java:628)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
at java.lang.Thread.run(Thread.java:745)
at com.hazelcast.util.executor.PoolExecutorThreadFactory$ManagedThread.run(PoolExecutorThreadFactory.java:58)
Caused by: com.hazelcast.spi.exception.TargetNotMemberException: Address : Address[10.174.36.244]:33002
at com.hazelcast.client.spi.impl.ClientInvocationServiceImpl.sendInternal(ClientInvocationServiceImpl.java:181)
at com.hazelcast.client.spi.impl.ClientInvocationServiceImpl.doSend(ClientInvocationServiceImpl.java:168)
at com.hazelcast.client.spi.impl.ClientInvocationServiceImpl.send(ClientInvocationServiceImpl.java:95)
at com.hazelcast.client.connection.nio.ClientConnection.heartBeatingSucceed(ClientConnection.java:396)
at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$HeartBeat.run(ClientConnectionManagerImpl.java:628)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
at java.lang.Thread.run(Thread.java:745)
at com.hazelcast.util.executor.PoolExecutorThreadFactory$ManagedThread.run(PoolExecutorThreadFactory.java:58)
at ------ End remote and begin local stack-trace ------.(Unknown Source)
at com.hazelcast.client.spi.impl.ClientCallFuture.resolveResponse(ClientCallFuture.java:182)
... 7 more

--- a/hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnection.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnection.java
@@ -396,8 +396,10 @@ public class ClientConnection implements Connection, Closeable {
     }
 
     void heartBeatingSucceed() {
-        if (failedHeartBeat != 0) {
-            if (failedHeartBeat >= connectionManager.maxFailedHeartbeatCount) {
+        final int lastFailedHeartBeat = failedHeartBeat;
+        failedHeartBeat = 0;
+        if (lastFailedHeartBeat != 0) {
+            if (lastFailedHeartBeat >= connectionManager.maxFailedHeartbeatCount) {
                 try {
                     final RemoveAllListeners request = new RemoveAllListeners();
                     final ICompletableFuture future = invocationService.send(request, ClientConnection.this);
@@ -406,7 +408,6 @@ public class ClientConnection implements Connection, Closeable {
                     logger.warning("Clearing listeners upon recovering from heart-attack failed", e);
                 }
             }
-            failedHeartBeat = 0;
         }
     }
 

